src/com/sun/jna/internal/Cleaner.java

```diff
@@ -151,4 +137,42 @@ void setNext(CleanerRef next) {
     public static interface Cleanable {
         public void clean();
     }
+
+    private class CleanerThread extends Thread {
+
+        private static final long CLEANER_LINGER_TIME = 30000;
+
+        public CleanerThread() {
+            super("JNA Cleaner");
+            setDaemon(true);
+        }
+
+        @Override
+        public void run() {
+            while (true) {
+                try {
+                    Reference<? extends Object> ref = referenceQueue.remove(CLEANER_LINGER_TIME);
+                    if (ref instanceof CleanerRef) {
+                        ((CleanerRef) ref).clean();
+                    } else if (ref == null) {
+                        synchronized (referenceQueue) {
+                            if (firstCleanable == null) {
+                                cleanerThread = null;
```

Comment:
I'm wondering if using an `ExecutorService` might make any of this easier.  See https://stackoverflow.com/questions/13883293/turning-an-executorservice-to-daemon-in-java